Refactor Human Resources and Payroll Management

jniles opened the pull request:
~~This PR is ready for testing and merge, once https://github.com/Third-Culture-Software/bhima/pull/7979 lands.~~

This PR is ready for testing and merging!  I recommend following the [guide for payroll here](https://docs.bhi.ma/fr/payroll/pratiques) to verify the functionality.

It's main contribution is a much better logical flow to the payroll code on the server, and address some of the UI complaints multiple_payroll module.  Because little was changed in the data model (SQL) or the core of the UI, the tests were not needed to be extensively updated.

The old PR description is found below to denote the journey we've been on in this PR.

# Previous PR Description:

This PR reworks the underlying Human Resources and Payroll Management code and procedures to allow for better code readability, maintainability, efficiency and testing.  It is part of an effort to rewrite the payment processes to support multiple payroll systems in a unified way (https://github.com/Third-Culture-Software/bhima/issues/7752).

All updates and changes are being made a feature branch `refactor-payroll-management`.  This is the PR pulls that branch into `master`.  It will be marked as a draft, and should not be merged until it is marked as needing review.

### General Approach

The general approach for this refactor is to:
 1. Improve code quality by using modern JS techniques (`async`/`await`, `Promise`, etc) and removing unused code paths.
 2. Add comments and `debug()` calls to complex tasks to document code behavior.
 3. Encapsulate common code within bespoke functions to make the code self-documenting (e.g. `isBenefitRubric()`)
 4. Server APIs gained the `payroll/` prefix where there might be confusion (e.g. `account_config/` -> `payroll/account_config/`)
 5. Tackle outstanding payroll issues from the issue tracker.

Some notable changes:
 1. The Rubric Configuration interface was unified (https://github.com/Third-Culture-Software/bhima/pull/7952).
 2. The Function UI gained additional descriptive text (https://github.com/Third-Culture-Software/bhima/pull/7955)
 
### Issues Tackled

The following issues are tacked in part or in whole:
 - https://github.com/Third-Culture-Software/bhima/issues/7791 (descriptive text only applies to the `commitmentByEmployee()` function so far).
 - https://github.com/Third-Culture-Software/bhima/issues/6097 (better English descriptions for payroll tasks).
 - https://github.com/Third-Culture-Software/bhima/issues/7797.
 -  https://github.com/Third-Culture-Software/bhima/issues/7821.
 - Closes https://github.com/Third-Culture-Software/bhima/issues/7966.

### Payroll Concepts and Definitions (as expressed by BHIMA)

For a review of the DRC Human Resources definitions, see https://github.com/Third-Culture-Software/bhima/issues/7969.  Below, we discuss how these concepts are expressed in BHIMA.

There are four basic types of payroll modifications: **benefits**, **payroll taxes**, **withholdings**, and **pension allowances**.

 -  **Benefits** are any rubrics that are paid to an employee.  In the current implementation of BHIMA payroll, the employee's base salary is classified as a benefit.  However, benefits can also be bonuses, additional compensation for family members, etc. 
     -  A benefits rubric is identified by the flag: `is_discount === 0`.
 - **Payroll Taxes** are called "charge sur remuneration" in French.  These taxes are paid by the enterprise and are not born by the employee.  
     - They are identified by the flags `is_employee !== 1`, `is_discount === 1`, `is_linked_pension_fund === 0`.  
 - **Withholdings** are monies retained/deducted by the institution from the employee's  pay.  This might be due to paying back salary advances, taking health insurance, or setting aside in an employee savings account.
     -  They are identified by `is_discount === 1` and `is_employee === 1`.
 - "**Pension Funds**" are retirement funds that are deducted from the employee salary.  
    - They are identified by `is_employee !== 0`, `is_discount === 1`, and `is_linked_pension_fund ===1`.

**Note**: we should probably rename "withholdings" to "deductions".

### Notable and/or Breaking Changes
 1. Renamed the API for `/account_config/` to `/payroll/account_config/` since it is used to configure accounts used in the payroll modules.
 6. Setting `DEBUG` to include `payroll:*` spews out much more informative information about the transaction process.
 7. The logic for determining different rubric types (e.g. taxes, pension, etc), has been moved to a `common.js` function and is shared across all commitment types.  It also has some rudimentary tests.
 8. `commitmentByEmployee()` now calls out to relevant `calculate*()` functions for taxes, withholdings, pension and benefits.  This should make the code somewhat more legible. 
 
 ### Testing 
We don't have very many payroll tests.  In particular, I'm not sure we have tests for all the different types of `commitment` functions listed in the enterprise settings.  Which means that, should this get merged, we cannot be 100% this code will not break something in production.

Ideally, I'm hoping to continue to hack on payroll over the holidays until we get something that is easy to reason about, maintain, and trace if a regression or a bug occurs.  However, we probably shouldn't do a release and/or upgrade until we have better test coverage or assurance that the payroll code isn't going to break in mysterious ways at one of the production sites.
 
 ### Future Work
 This PR addresses only about a quarter of the needed payroll refactor.  In many ways, it addresses the low-hanging fruit of function signatures and variable hoisting.  More in depth refactor is needed in the following areas:
  1. Add functions to determine taxable/nontaxable income to `payroll/multiplePayroll/common.js`, document them, and test them.
  4. Apply a similar refactor seen here to files and functions in `multiplePayrollIndice` associated with multiple indexes.
  6. Apply a similar refactor to `payrollSettings` and associated files.
  9. Rework the `commitFunction#dataCommit()` function to have a simpler function signature.
  10. Document the different payroll flows in the code.
  
On thing I noticed while working on this refactor: there are several places where we loop through employees and compute rubric values that are nearly identical.  For example, each `commitment` function  (`commitment.js`, `groupedCommitment.js`, `commitmentByEmployee.js`) loops through the employees, but generates slightly different transactions with the results.  With the current code structure, this is fine.  However, there is likely a way to make some sort of intermediate representation of the data that is common between all three that we can then reuse for all three.

As an example, one can imagine some sort of pseudocode like this:

```js
const employeeRubricAssignments = loopThroughEmployeesAndComputeRubrics(employees, rubrics) 

// if groupedCommitment.js
const transaction = createGroupedVouchers(employeeRubricAssignments);

// if commitment.js
const transaction =createCommitmentVouchers(employeeRubricAssignments);

// if commitmentByEmployee.js
const transaction = createIndividualVouchers(employeeRubricAssignments);
```

An advantage of this style is that we could then write unit tests for the `loopThroughEmployeesAndComputeRubrics()` function and ensure that we are correctly expressing the rubrics as transactions.  It would make it faster and easier to write a variety of tests for different scenarios (e.g. multiple pension withholdings, increases of greater than 100%, etc).

### Payroll Order of Operations
I spoke to Nancy Allan, who guided me through the basic payroll order of operations.  They are:
 1. Take the employee base salary.
 2. Compute the taxes due to CNSS (usually ~5%)
 3. Compute the taxes due to IPR (usually ~30%).
 4. Then you can add on housing and transport, which are typically untaxed.
 5. This provides the net salary that you will pay to the employee.
 6. Finally, the company/employer's taxes are calculated on the net salary.  This is usually ~13% of the gross salary.

It should be noted that pension and taxes on pension are separate calculations.

### Unanswered Questions:

1. How are payroll taxes that are born by employees, versus those born by the institution, modeled?
8. Why is there a different classification of rubrics in the "payroll by indice" modules?
9. How do we model overtime (heurs supplimentaire) using rubrics?  Are they automatically calculated?

jniles committed:
WIP: comment commitmentByEmployee()

This commit comments the commitmentByEmployee.js file as a work in
progress towards the major payroll refactor.

jniles committed:
WIP(payroll): add debug() calls.

jniles committed:
WIP(payroll): refactor/comment commitment.js

Adds detailed comments to the commitment.js file, as well as
refactoring the code in the following ways:
 1. Uses async/await in place of promises for better control flow.
 2. Use switch/case for explicit control flow.
 2. Delete unused variables, consolidate other variables.

jniles committed:
WIP(payroll): comment the dataCommitment function

Adds comments to the dataCommitment function.

jniles committed:
chore(multiplePayroll): add debug() comments

Adds debug comments for the multiple payroll configuration files.  Also
unwraps the function signature for setMultiConfiguration to no longer
take mysterious parameters.

jniles committed:
WIP(payroll): refactor payroll settings

Refactors the payroll settings controller to include additional
comments, debug() calls and reduce code complexity.

jniles committed:
WIP(payroll): refactor payroll account_config

Refactors and renames the API for payroll account configurations from
/account_config to /payroll/account_config to prevent confusion.
Updates comments to ensure correctness and usefulness.  Also migrates
the API to using async/await.

jniles committed:
chore: improve datelogic.js

Adds documentation comments.

jniles committed:
chore(commitment): fix houndci warnings

This commit adds the appropriate eslint skips to allow longer line
comments.

jniles committed:
WIP(payroll): improve commitments()

Adds comments, debug() calls, and refactors code to reduce side effects.
For example, many totally functions are consolidated into reduce()
functions.

jniles committed:
WIP(payroll): add description keys

Work-in-progress towards https://github.com/Third-Culture-Software/bhima/issues/7791.

Adds description keys to the commitmentByEmployee() function.

jniles committed:
WIP(payroll): refactor common fns into common.js

Moves rubric filtering and summing functions into a new common.js file
to help with legibility.

jniles committed:
chore(payroll): tighten up commitment.js

Tightens up the commitment.js file by removing the need for external
cost center code and using `const` to ensure that the arrays are not
reassigned.

jniles committed:
WIP(commitment): refactor commitment code

Refactors the commitment controller to more cleanly trace the payroll
process.

jniles committed:
refactor(cost center): remove assignCostCenter() function

Removes the assignCostCenter function that was previously in the cost
center module.  There is no need for it to be there, since it only deals
with array processing (though the array is a cost center).  Instead, it
is replaced with an anonymous function inline.

jniles committed:
test(payroll): add server unit test for common.js

Adds unit tests for the common.js modules.

jniles committed:
WIP(payroll): move cost center assignment fn to common.js

Moves the cost center assignment function to `payroll/common.js` since
it is shared between the groupedCommitment and commitment files.  Also
adds tests for all the common.js functions.

jniles committed:
refactor(payroll): optimize exchange rates with map

Reduces the complexity of exchange rate lookups across all commitment
functions by converting the exchange rates into an exchange rate map
with O(1) lookup time.  Reduces lookups from O(n^2) to O(n) lookups.server

jniles committed:
refactor(payroll): finalize clean up work

Finalizes the clean up work on payroll to provide a base for continued
improvement.

jniles committed:
WIP(payroll): rework commitment function signatures

Reworks the commitment function signatures to reduce the size of the
function signatures. Still a work in progress.

jniles committed:
WIP(payroll): simplify function call signature

This commit simplifies the function call signature by removing extra
parameters that are instead attached to the configuration object passed
into the subsequent commitment() calls.

jniles committed:
refactor(payroll): simplify commitmentByEmployee

Simplifies the commitmentByEmployee() function to make external calls
to newly implemented functions:

 1. calculateEmployeeBenefits
 2. calculateEmployeeWithholdings
 3. calculateEmployeePayrollTaxes
 4. calculateEmployeePension

 Each function filters the rubrics for their specific rubric type and
 accumulates the values by employee.

 The descriptions have now also been improved to record the rubric
 label, employee information, and payment period information to the
 voucher descriptive text.

 Closes #7791.

jniles committed:
chore(payroll): migrate setConfiguration() to async/await

Migrates the code that configures an employee for payment to async/await
to reduce verbosity and better control variable scope.

jniles committed:
chore(payroll): replace mkVoucher() with sharedVoucherProps

Replaces the function call with the object spread.

jniles committed:
chore(payroll): clean up comments and variable naming

Cleans up a few comments and inconsistent variable names noticed during
the review period.

jniles committed:
refactor(payroll): replace "date_embauche" with"hiring_date"

Replaces the `date_embauche` with `hiring_date` for better English
consistency.

Closes #7821.

jniles committed:
refactor(payroll): replace "date_embauche" with"hiring_date"

Replaces the `date_embauche` with `hiring_date` for better English
consistency.

Closes #7821.

jniles committed:
refactor(payroll): replace "date_embauche" with"hiring_date"

Replaces the `date_embauche` with `hiring_date` for better English
consistency.

Closes #7821.

jniles committed:
feat(bhCurrencyInput): add horizontal prop for form-horizontal

Enhances the `bhCurrencyInput` with a `horizontal` property that will
allow the currency to render correctly when on a form-horizontal form.

jniles committed:
chore(CurrencyService): update CurrencyService to modern ECMAScript

Replaces var with let/const in currency service and passes ESLint.

jniles committed:
chore(employees): rework employee registration page

Moves the rubrics for payroll lower in the "Optional Information"
section for better consistency.  It also cleans up the controller's
eslint issues.

jniles committed:
chore(HR): correct HR spelling in dictionary.

Corrects the spelling of "Human Resources" throughout the application.

jniles committed:
chore(rubrics): unify rubric configuration modals

Unifies the rubric configuration modal to configure both the label and
checkboxes for applied rubrics at the same time for a more streamlined
configuration.

jniles committed:
chore: add bold text

jniles committed:
chore(rubrics): unify rubric_config server controllers

Migrates the rubric configuration server controllers to use the
`payroll/` prefix in routes.  It also removes the `/setting` suffix from
the routes, opting instead to have an API similar to the cash payments
or invoices API where the items are provided as an array in the main
payment configuration object.

The next step will be to unify the client UI to allow simultaneous
configuration for the payroll rubrics.

jniles committed:
chore(payroll): Unify the interfaces for rubric configuration

Uses a unified interface for the rubric configuration service.

jniles committed:
chore(human resources): convert "fonctions" to async/await

Converts the server side code to async/await, removes duplicate
FunctionService.

jniles committed:
feat(functions): add column with num employees

Adds a column to the "function registry" to show the number of
employees with that function assigned.

jniles committed:
chore(accounts): rm unused chart-modal.html

Removes an unused, untranslated chart-modal.html.  No idea where it came
from.

jniles committed:
feat(functions): implement "batch create" for functions

Implements the same style of batch creates used on the account modal for
creating functions.

jniles committed:
chore(accounts): remove tech debt

Removes technical debt by removing duplicate code.

jniles committed:
fix(tax): rm race condition in TaxIPRConfigModal

Fixes a race condition that caused buggy behavior in the Tax IPR Config
Modal controller.  Now, the submit button is disabled until all the data
is available to use.

jniles committed:
fix(rubrics): do not crash on empty rubric items

Does not crash when the rubric items are empty.

jniles committed:
fix(tests): improve rubric tests

Block modal submission when required fields are not filled in and ensure
that `unset` tests work to allow creating a modal with empty rubric
fields.

jniles committed:
fix(employees): fix lint errors

Fixes lint errors in employee modules.

jniles committed:
chore(accounts): fix lint issues

jniles committed:
chore(payroll): address cosmetic requests

Addresses a number of cosmetic requests discussed in #6097.  Notably,
the English syntax for "nontaxable" and "payslip" have been addressed.
Additionally, the currency exchange rate is not rendered on the payslip
if the currency is the enterprise currency.  However, there is a bug with
the currency rendering on the payslip generator that does not allow the
currency to be properly rendered.  This will be addressed in a future
PR.

jniles committed:
perf(multiplePayroll): reference employee using UUID

Uses the employee's UUID to filter the employee registry from the
multiple payroll registry, rather than the human readable reference.

jniles committed:
chore(payroll): improve putOnWaiting modal

Improves the "putOnWaitingList" modal by using the modal itself to send
the HTTP request for the confirmation.  This allows the parent
controller to have a more standard flow.

Useless code was also removed, in addition to improving the variable
naming to better English names.

jniles committed:
refactor(payroll): improve config modal

Complete rewrite of the multiple_payroll payment configuration modal to
be less brittle and provide a better user experience.

I've completely removed the option for the user to set the currency
since no production sites were using it in practice, and it had some
serious bugs.  This reduces the number of imports into the modal and
removes the need to do any currency conversion.

Additionally, I've provided additionally context in the modal about the
employee, their grade, function, number of dependents, etc., to give the
user all the context needed for configuration of the employee.

Closes #7966.

jniles committed:
chore(payroll): improve UI of config modal

Improves the UI of the config modal with units for the different
days/people presented to the user.

jniles committed:
chore(deps): remove caniuse-lite

I am not sure when we added the canise-lite to the repository, but I
cannot find where it is used.  Removing for that reason.

jniles committed:
chore(payroll): whitespace fixes

Removes extra margin/padding to make the modal look nicer on smaller
screens.

jniles committed:
chore: lint fixes for payroll services

Runs `eslint --fix` on the services associated with human resources and
payroll.

jniles committed:
chore: bump deps

Bumps the dependencies.

jniles committed:
refactor: async/await server migration

Refactors more of the payroll code to async/await.  Merges the ipr_tax
services and prefixes the API routes with /payroll/taxes/ipr.

jmcameron commented on client/src/i18n/en/employee.json:
```diff
@@ -1,17 +1,18 @@
+    "TITLE": "Employee Management",
+    "NO_GRADE": "No grade assigned",
+    "VIEW_EMPLOYEE": "View Employee",
+    "CREATE_ANOTHER": "Create Another"
```
EMPLOYEE.CREATE_ANOTHER is missing from the French translation.
See `npm run check:lang`

jmcameron commented on client/src/js/services/CurrencyService.js:
```diff
@@ -16,10 +16,10 @@ CurrencyService.$inject = [
+  const service = this;
+  const baseUrl = '/currencies';
+  let cache;
+  let map;
```
👍

jmcameron commented on client/src/modules/employees/registration/employees.html:
```diff
@@ -1,16 +1,16 @@
   <div class="bhima-title">
     <ol class="headercrumb">
-      <li class="static" translate>TREE.HUMANS_RESSOURCES</li>      
+      <li class="static" translate>TREE.HUMAN_RESOURCES</li>      
```
👍

jmcameron commented on client/src/modules/employees/registration/employees.html:
```diff
@@ -292,7 +292,7 @@ <h4 translate>FORM.LABELS.CURRENT_LOCATION</h4>
                         type="text"
                         class="form-control"
-                        ng-model="EmployeeCtrl.employee.date_embauche"
+                        ng-model="EmployeeCtrl.employee.hiring_date"
```
👍

jmcameron commented on server/config/routes.js:
```diff
@@ -791,35 +789,35 @@ exports.configure = function configure(app) {
-  app.post('/iprTax', iprTax.create);
-  app.put('/iprTax/:id', iprTax.update);
-  app.delete('/iprTax/:id', iprTax.delete);
+  app.get('/payroll/taxes/ipr', iprTax.list);
```
👍

jmcameron commented on server/controllers/finance/cost_center.js:
```diff
@@ -295,33 +295,6 @@ function getAllCostCenterAccounts() {
   return db.exec(sql);
 }
 
-/**
```
Why was this removed?

jmcameron reviewed (changes requested):
I did a skimmy walk-through of the code changes.  There were too many to do a full evaluation.  But the changes and new code look good (as far as I could tell).

I found a translation term missing in French.

I need a better manual testing guide before completing the review.  

The basic and End-to-end tests all run fine on my testing machine.

Please merge with the  changes in master.  I could have done this from the github API, but I thought it would be better for you to do it.

jniles committed:
fix(commitmentByEmployee): properly format the txn

Fixes the transaction formatting for the commitmentByEmployee function
so that the db doesn't crash.

jniles committed:
fix(i18n): fix translated on errors messages

Updates the validation on inputs to correctly report "This field must be
a number" message if not a number is input.

jniles committed:
fix(payroll): improve commitment error guards

Expands the try/catch block of the commitment function to encompass more
errors.

jniles committed:
chore(i18n): fixup npm run check:lang

Fixes all the language warnings from our tooling.

jmcameron reviewed (approved):
My issues have been addressed.
The hands-on testing with @jniles went well.
The tests all pass locally.
